R/acs_dm.R

```diff
@@ -14,117 +14,133 @@ test_attr <- function(attr) {
-                  
-  if (x$geography == "tract"){
-    
-    df <- df %>% 
-      dplyr::rename(tract = GEOID) %>% 
-      dplyr::mutate(county_code = substr(tract, 1, 5))
-    
-    assertthat::assert_that(all(nchar(df$county_code) == 5))
-    assertthat::assert_that(length(unique(df$tract)) == nrow(df))
-    
-  } else if (x$geography == "zcta"){
-    
-    df <- df %>% 
-      dplyr::rename(zip = GEOID)
-    
-    assertthat::assert_that(length(unique(df$zip)) == nrow(df))
-    
-  }
   
-  assertthat::assert_that(all(df$n_kids_lt5 <= df$n_kids_lt6))
-  assertthat::assert_that(all(df$n_kids_lt4 <= df$n_kids_lt6))
-  assertthat::assert_that(all(df$n_kids_working_parents_lt5 <= df$n_kids_working_parents_lt6))
-  assertthat::assert_that(all(df$n_kids_working_parents_lt4 <= df$n_kids_working_parents_lt6))
-  assertthat::assert_that(all(df$n_kids_working_parents_lt6 <= df$n_kids_lt6))
+  assertthat::assert_that(all(x$df$n_kids_working_parents_lt5 <= x$df$n_kids_working_parents_lt6))
+  assertthat::assert_that(all(x$df$n_kids_working_parents_lt4 <= x$df$n_kids_working_parents_lt6))
+  assertthat::assert_that(all(x$df$n_kids_working_parents_lt6 <= x$df$n_kids_lt6))
 
-  return(df)
+  return(x)
 }
 
 #' @title Demand using table B17024
 #' @inheritParams dm
-dm.B17024 <- function(x) {
-
-  lt6 <- paste(x$table, "002", sep = "_")
-  lt6_under200_pct <- paste(x$table, c("003", "004", "005", "006",
-                                       "007", "008", "009", "010"), sep = "_")
-  
-  assertthat::assert_that(all(lt6_under200_pct %in% x$df$variable), msg = "Missing expected variables to create less than 200 pct")
+dm_acs.B17024 <- function(x) {
 
-  df <- x$df %>%
-    dplyr::select(GEOID, variable, estimate) %>%
-    dplyr::mutate(n_kids_lt6 = ifelse(variable %in% lt6, TRUE, FALSE),
-                  n_kids_lt6_under200pct = ifelse(variable %in% lt6_under200_pct, TRUE, FALSE)
-    ) %>%
-    tidyr::gather(variable2, value2, -c(GEOID, variable, estimate)) %>%
-    dplyr::filter(value2) %>%
-    dplyr::group_by(GEOID, variable2) %>%
-    dplyr::summarise(estimate = sum(estimate)) %>% 
-    tidyr::spread(variable2, estimate) %>%
-    dplyr::mutate(n_kids_lt5 = 5/6*n_kids_lt6,
+  x$df <- x$df %>%
+    dplyr::rename(n_kids_lt6_under200pct = n_kids_other_lt6) %>%
+    dplyr::mutate(n_kids_lt4_under200pct = 4/6*n_kids_lt6_under200pct,
                   n_kids_lt5_under200pct = 5/6*n_kids_lt6_under200pct,
-                  pct_kids_lt6_under200_pct = (n_kids_lt6_under200pct/n_kids_lt6)*100,
+                  pct_kids_lt4_under200_pct = (n_kids_lt4_under200pct/n_kids_lt4)*100,
                   pct_kids_lt5_under200_pct = (n_kids_lt5_under200pct/n_kids_lt5)*100,
-                  n_kids_lt4 = 4/6*n_kids_lt6,
-                  n_kids_lt4_under200pct = 4/6*n_kids_lt6_under200pct,
-                  pct_kids_lt4_under200_pct = (n_kids_lt4_under200pct/n_kids_lt4)*100)
-                  
-  if (x$geography == "tract") {
```

Comment:
I changed this to use classes instead of ifelse. Assigning classes to object then calling the parent function is a bit more tidy than the ifelse. Additionally, changing these steps to functions allows us to reuse the code for each ACS table. Let me know if you have any questions on the classes part of the code. They're a bit confusing at first, but once you understand them they can make coding much neater than the traditional if{}else{}